WIP: Remove use of SIMPL_* macros and replace with modern C++ code

imikejackson opened the pull request.

imikejackson committed:
Wrap the Python wrapping macros in a #define for SIMPL_ENABLE_PYTHON

This is done because SIMPLibSetGetMacros is going away eventually and we need to separate
out the still needed macros that help generate the python bindings.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Massive code modifications to clean up and reduce macro clutter

* Remove use of all SIMPL_* macros replacing with actual C++ code
* Remove the include directive to DataContainerArray in AbstractFilter instead using a forward declaration and alias
* Add needed include directives or forward declarations to all effected classes

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Transfer some SIMPL_* macros into the Pybind11 macro section

This is needed because the codescraper algorithm looks for four specific macros in order
to determine the proper c++ and python code to generate. The same macros have been
defined in the SIMPLPythonMacros.h header file.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix pass by const reference for POD types

In the last pass at removing the macros POD types were mistakenly passed by
const reference instead of by value. These have all been fixed to pass by value.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/Plugins/OrientationAnalysis/Gui/FilterParameterWidgets/EnsembleInfoCreationWidget.cpp:
```diff
@@ -1,4 +1,6 @@
 /* ============================================================================
+
+
```
Unintentional spacing

mmarineBlueQuartz reviewed (commented):
`const T&`
- int
- bool
- float
- double
- int64_t
- int32_t
- uint32_t
- uint64_t
- size_t

imikejackson committed:
Resolving merge conflicts

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Removing the macros is the default tool

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Merge OrientationTransformation and supporting classes API changes

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix warnings about returning const QString as a value instead of a reference.

Warnings are fixed by changing from "const QString" to "QString"

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix compiler warnings about overriding a virtual method without using override keyword

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix DOxygen comments that got munged up in some find/replace operation

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix compiler errors on MSVC platform

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

mmarineBlueQuartz commented on Source/OrientationLib/Utilities/ComputeStereographicProjection.h:
```diff
@@ -32,7 +32,7 @@
 
 #include "SIMPLib/DataArrays/DataArray.hpp"
-
+#include "SIMPLib/DataArrays/DataArray.hpp"
```
This is included twice

mmarineBlueQuartz commented on Source/OrientationLib/Utilities/ModifiedLambertProjection.h:
```diff
@@ -37,7 +37,7 @@
 
 #include "SIMPLib/SIMPLib.h"
-#include "SIMPLib/Common/SIMPLibSetGetMacros.h"
+#include "SIMPLib/DataArrays/DataArray.hpp"
 #include "SIMPLib/DataArrays/DataArray.hpp"
```
DataArray.hpp is included twice

mmarineBlueQuartz commented on Source/OrientationLib/Utilities/PoleFigureUtilities.h:
```diff
@@ -37,6 +37,7 @@
 
 #include "SIMPLib/SIMPLib.h"
 #include "SIMPLib/DataArrays/DataArray.hpp"
+#include "SIMPLib/DataArrays/DataArray.hpp"
```
Double include on DataArray.hpp

mmarineBlueQuartz commented on Source/Plugins/EMMPM/EMMPMLib/Core/EMMPM.h:
```diff
@@ -48,28 +47,76 @@
-    SIMPL_STATIC_NEW_MACRO(EMMPM);
-    SIMPL_TYPE_MACRO(EMMPM);
+    using Self = EMMPM;
+    ;
```
There is a random semicolon here, as well as one after `Q_OBJECT;`

imikejackson committed:
Update codes to latest const correctness changes

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Add missing include directive <memory> for shared_ptr<> access.

Updated some other code areas with better documentation strings
Updated some code sections to be more consistent with use of type aliases.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson committed:
Fix bug using std::accumulate due to casting issue.

Passing in "1" will force the computation to be done in int32_t. When using large
Tuple dimensions we will over flow.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>
